### ISSUE TITLE
NodeJS Device Client SDK throws exception after 45 minutes

### ISSUE THREAD

**Issue by robertwesterlund** [Azure/azure-iot-sdks#958]
When using the azure-iot-device@1.1.0-dtpreview and the azure-iot-device-mqtt@1.1.0-dtpreview npm packages to just open a client and then get a device twin, the libraries throw an exception after 45 minutes (during two consecutive tests it was exactly 45 minutes both times).

The relevant code part is only this:
``` 
process.on('uncaughtException', function(e){
	console.error('Uncaught exception', e);
	process.exit(-1);
});

const Client = require('azure-iot-device').Client;
const Protocol = require('azure-iot-device-mqtt').Mqtt;
const iotHubConnectionString = process.argv[2]; 
if (!iotHubConnectionString){
	console.log('Usage: node <jsfile>.js "<iothubconnectionstring>"');
}

let client = Client.fromConnectionString(iotHubConnectionString, Protocol);
client.open(function(error){
	if (error){
		console.log('Could not connect the client', error);
		process.exit(-2);
	}
	client.getTwin(function(error, twin){
		if (error){
			console.log('Could not get the twin', error);
			process.exit(-3);
		}
		console.log('Retrieved twin, waiting a looong time for the error');
	});
});
``` 

And the error received after 45 minutes is the following:

``` 
Error: this.mqttprovider.connect is not a function
    at Mqtt.connect (C:\git\iot-test\node_modules\azure-iot-mqtt-base\lib\mqtt.js:74:35)
    at Mqtt.connect (C:\git\iot-test\node_modules\azure-iot-device-mqtt\lib\mqtt.js:49:14)
    at Client.open (C:\git\iot-test\node_modules\azure-iot-device\lib\client.js:323:21)
    at Client.<anonymous> (C:\git\iot-test\node_modules\azure-iot-device\lib\client.js:283:14)
    at Mqtt.<anonymous> (C:\git\iot-test\node_modules\azure-iot-device-mqtt\lib\mqtt.js:148:7)
    at Store.close (C:\git\iot-test\node_modules\mqtt\lib\store.js:104:5)
    at C:\git\iot-test\node_modules\mqtt\lib\client.js:521:26
    at Store.close (C:\git\iot-test\node_modules\mqtt\lib\store.js:104:5)
    at Immediate.closeStores (C:\git\iot-test\node_modules\mqtt\lib\client.js:520:24)
    at runCallback (timers.js:576:20)
``` 

The exact test script used to reproduce the issue is attached as a zip file with the js file, the package.json file and the yarn.lock file.

[iot-test.zip](https://github.com/Azure/azure-iot-sdks/files/592744/iot-test.zip)

**Comment by robertwesterlund** [comment]
I'm not sure if it will turn out to be relevant information, but the above was verified on more than one computer, but only on node v6.7.0.

**Comment by BertKleewein** [comment]
Hi Robert,  

That exact stack trace isn't familiar to me, but the fact that it's exactly 45 minutes leads me to believe that this is a problem renewing the SAS token.  I put in a fix after the 1.1.0-dtpreview drop.  The fix is here: https://github.com/Azure/azure-iot-sdks/commit/aee4e2124287f1891575e02652834b8a70005ecf 

If you look at the commit delta, you can see comments about renewing every 45 minutes.  This interval didn't change with the fix, but the handling of the renewal did.  If you want to change the interval to shorten your verification time, you can lower the value of Client.sasRenewalInterval

You can get this fix in azure-iot-device@1.1.0-dtpreview.2 or in azure-iot-device@1.0.17.  There is also a new version of this library coming "any day now" that you may want to watch for.

**Comment by robertwesterlund** [comment]
Great, the SAS renewal fixed the issue. However, even though the SAS renewal now works and client can keep sending things, after a SAS renewal calling a device client method now fails.  Changing the SAS renewal time changes the period length that it works, so it seems very much to be related. I have logged issue #966 for this.
